quinn/tests/many_connections.rs

```diff
@@ -0,0 +1,203 @@
+    /// finished.
+    received_messages: usize,
+    expected_messages: usize,
+
+    /// Hold connections so that streams wouldn't be closed prematurely
+    connections: Vec<quinn::Connection>,
+}
+
+impl TestContext {
+    /// Constructs test context and returns test completion receiver as well..
+    fn shared() -> Rc<RefCell<Self>> {
+        Rc::new(RefCell::new(Self {
+            received_messages: 0,
+            expected_messages: 0,
+            connections: Default::default(),
+        }))
+    }
+}
+
+#[test]
+#[ignore]
+fn connect_n_nodes_to_1_and_send_1mb_data() {
+    let mut runtime = unwrap!(Runtime::new());
+
+    let (cfg, listener_cert) = configure_listener();
+    let mut ep_builder = quinn::Endpoint::new();
+    ep_builder.listen(cfg);
+    let (driver, endpoint, incoming_conns) = unwrap!(ep_builder.bind(&("127.0.0.1", 0)));
+    runtime.spawn(driver.map_err(|e| panic!("Listener IO error: {}", e)));
+    let listener_addr = unwrap!(endpoint.local_addr());
+
+    let ctx = TestContext::shared();
+    ctx.borrow_mut().expected_messages = 50;
+
+    let ctx2 = ctx.clone();
+    let read_incoming_data = incoming_conns
+        .map_err(|()| panic!("Listener failed"))
+        .for_each(move |(conn_driver, conn, incoming)| {
+            current_thread::spawn(conn_driver.map_err(|_| ()));
+
+            let ctx = ctx2.clone();
+            let task = incoming
+                .map_err(move |e| panic!("Incoming streams failed: {}", e))
+                .for_each(move |stream| {
+                    read_from_peer(stream, ctx.clone(), conn.clone());
+                    Ok(())
+                })
+                .then(move |_| Ok(()));
+            current_thread::spawn(task);
+
+            Ok(())
+        });
+    runtime.spawn(read_incoming_data);
+
+    let client_cfg = configure_connector(&listener_cert);
+    let (conn_done_tx, conn_done_rx) = mpsc::unbounded();
+
+    for _ in 0..ctx.borrow_mut().expected_messages {
+        let data = random_data_with_hash(1024 * 1024);
+        let ctx = ctx.clone();
+        let conn_done_tx = conn_done_tx.clone();
+
+        let task = unwrap!(endpoint.connect_with(&client_cfg, &listener_addr, "Test"))
+            .map_err(|e| panic!("Connection failed: {}", e))
+            .and_then(move |(conn_driver, conn, _)| {
+                current_thread::spawn(conn_driver.then(move |_| {
+                    // connection was closed, notify about it
+                    let _ = unwrap!(conn_done_tx.unbounded_send(()));
+                    Ok(())
+                }));
+                write_to_peer(&conn, data);
+                ctx.borrow_mut().connections.push(conn);
+                Ok(())
+            });
+        runtime.spawn(task);
+    }
+
+    let wait_for_all_connections = conn_done_rx
+        .take(ctx.borrow().expected_messages as u64)
+        .for_each(|_| Ok(()));
+    unwrap!(runtime.block_on(wait_for_all_connections));
```

Comment:
Could [`Runtime::run`](https://docs.rs/tokio/0.1.18/tokio/runtime/current_thread/struct.Runtime.html#method.run) be used here instead of explicit message-passing?

Reply:
It's a neat idea, thanks. Although I will still have to notify endpoint driver that I'm done with the test, cause it is the last spawned task blocking `Runtime::run()` :)

Reply:
Could you accomplish that without explicit messaging by dropping the `Incoming` stream (along with all other `Endpoint` handles) after the expected number of connections have been handled?

Either way, it's nice to invoke `run` here to ensure that there aren't any stale futures left floating around, since the shutdown logic is nontrivial.

Reply:
hm, `Incoming` and `EndpointDriver` are spawned on the event loop, so I don't have handles to them anymore. Besides signaling them with messaging, I don't know how I could drop those tasks :/

Reply:
You know how many connections you're expecting, so you can use `Stream::take` to finish the `Incoming` stream after that number; once the resulting future is finished, the runtime will automatically drop it. `EndpointDriver` will automatically finish (and be dropped) once `Incoming` and all `Endpoint` values have been dropped and all `Connection`s are complete.